### PR TITLE
gpl: limit filler cell count on low density designs

### PR DESCRIPTION
**Pull request by gadfort**
Issue:
- on large designs with few place-able instances the number of fill cell added during global placement can be huge and significantly slow down placement (in my case ~50000 instances with ~25M filler cells).

Changes:
- during filler cell initialization this will increase the filler cell sizes if the filler area is 10x the place-able cell area. This limits the the number of filler cells to ~10x the number of instances or ~1M.
- On designs with reasonable densities this code will not get triggered.

With this change the previously mentioned placement ends up with ~500k filler cells and placement ends in 5 minutes instead of 100 minutes.

The constants were chosen based on a single testcase, so we should determine what makes sense here (provide API access to these or different tuning?).

### CODE REVIEW

**Review by maliberty — CHANGES_REQUESTED** [review]

File: src/gpl/src/nesterovBase.cpp [hunk 1]
```diff
@@ -1434,10 +1436,52 @@ void NesterovBase::initFillerGCells()
                 uniformTargetDensity_);
   }
 
-  int fillerCnt = static_cast<int>(
+  // limit filler cells
+  static constexpr double limit_filler_ratio = 10;
+  const double filler_scale_factor = std::sqrt(
+      totalFillerArea_ / (limit_filler_ratio * nesterovInstanceArea));
+  if (filler_scale_factor >= 1.0) {
```
[the remark this answers is not in the document]
Comment (gadfort) [reply src/gpl/src/nesterovBase.cpp]: that correct, I removed it

**Comment by maliberty** [comment]
Started a secure CI - I don't expect any designs to have a low enough util for this to matter but making sure.

**Comment by stefanottili** [comment]
Why does it take 100min to add 25M fillercells?

**Comment by gadfort** [comment]
@stefanottili it takes 100mins to place 25M cells

**Comment by stefanottili** [comment]
Sorry, now I'm confused. You mentioned
> ... on large designs with few place-able instances the number of fill cell added during global placement can be huge and significantly slow down placement (in my case ~50000 instances with ~25M filler cells).
> With this change the previously mentioned placement ends up with ~500k filler cells and placement ends in 5 minutes instead of 100 minutes.

I read this as "large floorpan, 50k cells, 25M filler cells in 100min", after change
"... filler cell initialization .. will increase the filler cell sizes: large floorpan, 50k cells, 500k large filler cells in 5min"

And I was wondering
a) why does it take 100s to add 25M filler cells and also
b) do you only generate abstracts for the "increased fill cell sizes" or also layout ?

I don't get your answer to a) You're saying
> it takes 100min to place 25M cells

25M stdcells or 25M filler cells ?
"place" .. are you referring to global placement (gpl), detailed placement (dpl) or filler cell insertion ?

**Comment by gadfort** [comment]
@stefanottili the filler cells here are the internal cells of the global placer (they are virtual only, nothing real about them), it does not relate to the detailed placer or filler insertion.
The filler cells are added by GPL to fill up the empty area on the design. In my case, it's a very sparse design so it's mostly empty, thus causing GPL to add a lot of fillers (internal cells). Effectively each of these needs to be placed by the global placer even though they are not real, which ends up taking a lot of time. By increasing the effective size of these filler cells we get fewer of these and thus the algorithm doesn't take a such time.

**Comment by stefanottili** [comment]
@gadfort thanks for the explanation, highly appreciated

**Comment by maliberty** [comment]
All public tests pass.  The smaller private ones have as well so I'm ok to go ahead with this (I don't believe an large ones are this sparse).

**Review by maliberty — APPROVED** [review]

**Pull request merged**